### ISSUE TITLE
Zodiac GX

### ISSUE THREAD

**Issue by Dru90** [faucetsdn/faucet#2349]
Hi,

I'm currently attempting to setup Faucet + Gauge + Prometheus + Grafana using a Zodiac GX.

Following the documentation below:

"ZodiacGX is a small 5 port multi table OF1.3 switch from Northbound Networks. Please see the documentation for configuring OpenFlow on the switch, and use ZodiacGX as the FAUCET hardware type."

Faucet doesn't recognise Zodiac GX as a hardware type - only Zodiac FX will work when running check_faucet_config. I attempted to use Zodiac FX as a hardware type for the Zodiac GX but I could not get the switch to work using openflow.

Any advice? Cheers

**Comment by Dru90** [comment]
- I'm refering to the hardware type you set in the faucet.yaml file

**Comment by shivarammysore** [comment]
Use the latest version of Faucet as I saw a fix go in a short while ago 

/Shivaram
::Sent from my mobile device::

**Comment by Dru90** [comment]
Will do thanks. 

Is there a template faucet.yaml file for the Zodiac GX?

**Comment by Dru90** [comment]
$ check_faucet_config /etc/faucet/faucet.yaml 
hardware ZodiacGX must be in ['NoviFlow', 'CiscoC9K', 'GenericTFM', 'Open vSwitch', 'Aruba', 'Lagopus', 'Netronome', 'Allied-Telesis', 'ZodiacFX']

Can confirm same problem on latest version

**Comment by shivarammysore** [comment]
Scroll to end of our this file 
https://github.com/faucetsdn/faucet/blob/master/faucet/valve.py or search for Zodiac 
If it is not this release, you can download master and build binaries yourself or wait a week for next release

/Shivaram
::Sent from my mobile device::

**Comment by byllyfish** [comment]
According to the docs, the Zodiac GX defaults to making OF connections to port 6633.  While Faucet itself may listen on both 6653 and 6633, the Dockerfile only exports 6653. The ZodiacGX docs look like they expose the OVS settings that would allow you to change the port to 6653. Can you try using port 6653?

Using 'ZodiacFX' or 'Open vSwitch' in faucet.yaml should work; it aliases the Open vSwitch setting.

**Comment by gizmoguy** [comment]
Hi @Dru90,

ZodiacGX was introduced as a hardware type recently in faucet@master and it hasn't been included in a faucet release yet. It will be included in 1.8.17 which is due to be released in 3 days from now.

Luckily for you hardware types are just informational for faucet so for now you can just use "Open vSwitch" as your hardware type and it will work fine.

In terms of a skeleton faucet configuration, here is the configuration I am using to drive mine on my production network:

https://github.com/wandsdn/redcables-ansible/blob/testing/files/faucet/l2/access.yaml#L697-L720

What I did was remove eth0.5 from the OVS bridge and I'm using it as an out of band control plane network port, however as long as you remember to include the VLAN your OpenFlow controller is on as an native_vlan on one of the ports to the OVS bridge on the ZodiacGX then in band control should also work.

Here is my current OVS config from my ZodiacGX:

```
admin@ZodiacGX:~$ sudo ovs-vsctl show
f6d3dd59-556b-4648-b4e0-af7f2773119f
    Bridge ovslan
        Controller "tcp:172.16.0.201:6654"
            is_connected: true
        Controller "tcp:172.16.0.200:6653"
            is_connected: true
        fail_mode: secure
        Port ovslan
            Interface ovslan
                type: internal
        Port "eth0.3"
            Interface "eth0.3"
        Port "eth0.1"
            Interface "eth0.1"
        Port "eth0.4"
            Interface "eth0.4"
        Port "eth0.2"
            Interface "eth0.2"
```

To get to this point what I did was delete all the code from the startup script (/etc/rc.local) leaving just "exit 0" at the end, OVS has a persistent database (ovsdb) so we don't need to run this every boot.

I then manually tweaked my OVS bridge configuration:
```

admin@ZodiacGX:~$ ovs-vsctl set bridge ovslan stp_enable=false
admin@ZodiacGX:~$ ovs-vsctl set-fail-mode ovslan secure
admin@ZodiacGX:~$ ovs-vsctl set-controller ovslan tcp:172.16.0.200:6653 tcp:172.16.0.200:6654
```

**Comment by Dru90** [comment]
@gizmoguy thanks for the detailed reply. I like the idea of having an out of band control plane port.

I'll attempt to implement your configuration at some point today.

**Comment by Dru90** [comment]
@gizmoguy how do you remove eth0.5 from the OVS Bridge?

**Comment by shivarammysore** [comment]
ovs-vsctl del-port bridge_name

Please check OVS-VSCTL documentation for exact syntax and usage

**Comment by Dru90** [comment]
@shivarammysore thanks

**Comment by Dru90** [comment]
I've now removed eth0.5 from the OVS bridge. 

At this point how would I configure and access port eth0.5 out of band?

**Comment by shivarammysore** [comment]
Connect that to the TOR switch/router.  Faucet automatically recognizes the
same :-)

**Comment by Dru90** [comment]
I've successfully setup faucet gauge prometheus and grafana!

One last question :) - The faucet_inventory.json file that you import into Grafana displays a dashboard that lists the Zodiac GX twice... Is this normal behaviour? If so why does it do this? Screenshot attached
![2018-08-25-223649_1824x984_scrot](https://user-images.githubusercontent.com/42614550/44622781-f9e7ce00-a8b7-11e8-81cb-2642ee6da23e.png)

**Comment by Dru90** [comment]
@gizmoguy Ok what I thought was working actually wasn't. The datapath was lost after about 60 seconds. 

What I did: 

1. deleted port eth0.5 from the OVS bridge
2. assigned an IP address to eth0.5 (ifconfig eth0.5 172.16.0.15) through the CLI of the ZodiacGX that was on the same network as the controller (172.16.0.20).
3. The datapath was established (as can be seen in screenshot above) for roughly 60 seconds. Then the connection drops.....

Am I going about this wrong?

**Comment by gizmoguy** [comment]
Setting ``connection-mode`` to out-of-band removes the default OVS "hidden flows" which enable it to find an openflow controller in-band. Please read the "_Q: Some of the traffic that I’d expect my OpenFlow controller to see doesn’t actually appear through the OpenFlow connection, even though I know that it’s going through._" question on http://docs.openvswitch.org/en/latest/faq/openflow/ for more details.

It looks like grafana's display code for tables has changed recently. I have pushed a new json dashboard in #2353 which is now published here: https://docs.faucet.nz/en/latest/_static/grafana-dashboards/faucet_inventory.json. Please download a reimport into grafana.

Not sure why your control plane connection drops every 60 seconds. Could you please attach the contents of /var/log/faucet/faucet.log from your faucet controller and provide the output from running "sudo logread | grep ovs-vswitchd" on the zodiacgx.

**Comment by gizmoguy** [comment]
Yes we have meter support in faucet, documentation is here:

https://faucet.readthedocs.io/en/latest/configuration.html#configuration-meters

and a configuration example is here:

https://github.com/faucetsdn/faucet/blob/edbb672b35e65b0729e238fac4096000cec877d0/tests/integration/mininet_tests.py#L590-L610

Unfortunately the ZodiacGX does not come with meter support. It is build using OVS 2.5.0 and meters were added to OVS recently in version 2.10.0.

**Comment by Dru90** [comment]
Is there an alternative way to implement QOS on the Zodiac GX using Faucet?

**Comment by gizmoguy** [comment]
You could NFV your packets with an ACL to a separate host that is running TC to implement QoS. Or possibly you could add TC to the ZodiacGX since it does everything in software?

Brad

**Comment by anarkiwi** [comment]
It looks to me like all FAUCET related issues have been addressed here (and we have just released 1.8.17 which adds the ZodiacGX hardware type. If you find any other issues with FAUCET please do open a new issue.

**Comment by Dru90** [comment]
Yeah no worries. Will do thanks
